### PR TITLE
Separate an SQSStream from MessageStream

### PR DESCRIPTION
**Pull request by alexwlchan**
### What is this PR trying to achieve?

Currently we have a MessageStream class, which streams message pointers from SQS and fetches the associated object from S3.

This PR pulls out the SQS streaming logic, so we can reuse it in other applications where we want to read from SQS with backpressure. It's mostly a copy-and-paste of code from MessageStream into a new class, with tests likewise copied and tweaked only slightly.

### Who is this change for?

Me, as I want to make the Sierra readers backpressure-aware.

### CODE REVIEW

File: sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/sqs/SQSStream.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        metricsSender.incrementCount(
+          metricName = s"${streamName}_MessageProcessingFailure",
+          count = 1.0
+        )
+    }
+    processMessageFuture
+  }
+
+  protected def read(message: sqs.model.Message)(implicit decoderT: Decoder[T]): Future[T] =
+    Future.fromTry(fromJson[T](message.getBody))
```
Comment (alexwlchan) [inline sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/sqs/SQSStream.scala]: This is a slight case of the tail wagging the dog – I know we need to override this with a method that returns a Future below, so it’s a Future here to avoid the complexity of modifying readAndProcess() to support a Future or a Try. In an ideal world it would just return a Try, but I don't think it's that important.

File: sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/message/MessageStream.scala [hunk 2]
```diff
@@ -26,60 +19,19 @@ class MessageStream[T] @Inject()(actorSystem: ActorSystem,
...
-          1.0)
-    }
-    processMessageFuture
-  }
-
-  private def read(message: sqs.model.Message)(
-    implicit decoderN: Decoder[NotificationMessage],
-    decoderT: Decoder[T]
-  ): Future[T] = {
+  override def read(message: sqs.model.Message)(implicit decoderT: Decoder[T]): Future[T] = {
```
Comment (alicefuzier) [inline sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/message/MessageStream.scala]: Uhm, one of the things I really don't like in `MessageWorker` and `SqsWorker` is using polymorphism rather than composition. It makes it hard to follow how those functions are used. I'd rather not introduce the same pattern here.
Because there is no mutable state there should be a way of just passing functions around.

Comment (alexwlchan) [reply sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/message/MessageStream.scala]: So instead you have to pass a `read()` method into `foreach`? Is that the idea?

Comment (alicefuzier) [reply sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/message/MessageStream.scala]: To further extend what I mean, I think `MessageStream` should be using `SQSStream` rather than extending it. Its `foreach` function should internally call `SQSStream.foreach` passing it a function that goes from `NotificationMessage` to `Future[Unit]`. That function should be in turn be made of:
- decoding the `MessagePointer` and getting the object from s3
- decoding the object from s3 into T
- apply the `process` function supplied by the user in `MessageStream.foreach`

Comment (alexwlchan) [reply sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/message/MessageStream.scala]: Ah, got it. Yeah, that makes sense.

File: sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/sqs/SQSStream.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        logger.error(s"Failure while processing message.", exception)
+        metricsSender.incrementCount(
+          metricName = s"${streamName}_MessageProcessingFailure",
+          count = 1.0
+        )
+    }
+    processMessageFuture
+  }
+
+  protected def read(message: sqs.model.Message)(implicit decoderT: Decoder[T]): Try[T] =
```
Comment (alicefuzier) [inline sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/sqs/SQSStream.scala]: This can be private now

**Comment by alicefuzier** [comment]
Yep, I like this 👍
